MBS-4685 / MBS-13084: Allow modifying and removing edit notes

reosarevok opened the pull request:
### Implement MBS-4685 / MBS-13084

# Problem
There is no way for users to delete or fix their edit notes. While obviously this is not desirable all the time, it would be great to have a way for editors to delete or change an edit note if they, say, immediately notice they pasted the wrong thing (even more so if it's for example private data).

There's also no way for *admins* to delete or amend edit notes, which means notes with spam such as [this one](https://musicbrainz.org/edit/96631022#note-96631022-1) cannot be acted on.

# Solution
This adds a new table `edit_note_change` to track changes to edit notes.

Admins can edit or delete an edit note at any time of their choosing. For anyone else, the following conditions must apply:
1) the edit note is by the user (obviously)
2) the edit note hasn't already been deleted
3) the edit note isn't older than 24 hours
4) the edit note hasn't gotten a reply by anyone else than the editor or ModBot
5) the user isn’t disabled to add edit notes (by admin)

If the edit note is too old or has a reply, but the user really wants to edit it (say, someone answered to say "Is that your password? Whoops!"), then we tell them to contact us so an admin can make the necessary changes if they agree with the reasoning.

![Screenshot from 2021-12-08 16-39-22](https://user-images.githubusercontent.com/1069224/145383826-973143fe-efc7-40a0-85a6-91a0a8f40043.png)

# Testing
Manually, plus wrote a few tests.

Tested manually: 
* Added notes to stuff, tried removing them with a reason, without a reason, editing with a reason, without a reason, editing and deleting them as admin.

Covered in tests:
* The expected amount of modify note controls are shown for a non-admin editor and for an admin (so, the expected amount of notes are editable).
* Edit notes cannot be modified by non-admin non-owner even by navigating to the form directly.
* Edit notes with replies cannot be modified by non-admin owner even by navigating to the form directly.
* Old edit notes cannot be modified by non-admin owner even by navigating to the form directly (limit difference is 24 hours, but tested difference is many years).
* Edit notes cannot be modified by non-admin owner when their edit note privs are off, even by navigating to the form directly.
* Edit owner can modify edit note even if themselves and ModBot have left further notes, as long as nobody else has.
* Actually modifying the edit note as the owner works, and after modification the reason and date of the modification are listed.
* Actually deleting the edit note as the owner works, and after deletion the reason of the deletion is listed.
* Admins can delete old edit notes.
* Admins can "delete" already deleted edit notes to change the reason (for example if the deleting editor left an insult there).
* Actually deleting the edit note as an admin works, and after deletion the reason of the deletion (or the fact there was no reason) is listed.
* Edit notes cannot be edited to a blank string, or a string devoid of content like "a" (as with required edit note additions).
* Already deleted edit notes cannot be deleted again nor edited by their owner.

reosarevok commented:
@brainzbot, retest this please

mwiencek commented on admin/sql/CreateTables.sql:
```diff
@@ -576,6 +576,20 @@ CREATE TABLE edit_note
     post_time            TIMESTAMP WITH TIME ZONE DEFAULT NOW()
 );
 
+CREATE TYPE edit_note_status AS ENUM ('deleted', 'edited');
```
We actually have admin/sql/CreateTypes.sql now which this should probably be moved to

mwiencek commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,208 @@
+          component_path  => 'user/UserMessage',
+          component_props => {
+              title    => l('Cannot modify edit note'),
+              message  => l('You cannot modify other users’ edit notes'),
```
(Mainly added the period, but another message below uses `can’t`, so suggesting that too for consistency.)
```suggestion
              message  => l('You can’t modify other users’ edit notes.'),
```

mwiencek commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,208 @@
+          component_path  => 'user/UserMessage',
+          component_props => {
+              title    => l('Cannot modify edit note'),
+              message  => l('You can’t modify this, since somebody else has already replied to it.'),
```
```suggestion
              message  => l('You can’t modify this note, since somebody else has already replied to it.'),
```

mwiencek commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    my $all_edit_notes = $edit->{edit_notes};
+    my $note_index = first {
+      @$all_edit_notes[$_]->id == $edit_note->id
+    } 0..@$all_edit_notes;
```
This might be handy here https://metacpan.org/pod/List::AllUtils#first_index-BLOCK-LIST
```suggestion
    my $note_index = first_index {
      $all_edit_notes->[$_]->id == $edit_note->id
    } @$all_edit_notes;
```

reosarevok replied:
Guess you meant just `$_->id == $edit_note->id` for the second line?

mwiencek commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,208 @@
+    my $note_index = first {
+      @$all_edit_notes[$_]->id == $edit_note->id
+    } 0..@$all_edit_notes;
+    my @after_note = splice(@$all_edit_notes, $note_index + 1);
+    my $has_reply = 0;
+    for (@after_note) {
+      if ($_->editor_id != $edit_note->editor_id){
+        $has_reply = 1;
+      }
+    }
```
I'd suggest just using a for loop (instead of splicing elements from the original array).  It looks like you can also break early if a reply is found.
```suggestion
    my $has_reply = 0;
    for (my $i = $note_index + 1; $i < @$all_edit_notes; $i++) {
      if ($all_edit_notes->[$i]->editor_id != $edit_note->editor_id) {
        $has_reply = 1;
        last;
      }
    }
```

mwiencek commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,208 @@
+    $self->detach_if_cannot_modify($c, $edit_note, $edit);
+
+    if ($c->form_posted_and_valid($form)) {
+        if ($form->field('cancel')->input) {
+          # Do nothing
+        } else {
```
Do you prefer the empty condition in this case or?
```suggestion
        if (!$form->field('cancel')->input) {
```

reosarevok replied:
Not necessarily, although maybe it's a bit more clear? I just saw everywhere else we have this `if cancel then x else run` construction so it seemed more consistent, but if you prefer the other way I can change it.

mwiencek commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,208 @@
+        } else {
+          $c->model('MB')->with_transaction(sub {
+            $c->model('EditNote')->remove_content(
+              $edit_note,
```
A lot of the indentation in this file is two spaces instead of four

reosarevok replied:
Oh, can change. That said, is there a reason we generally do 2 spaces in JS and 4 in Perl?

mwiencek commented on lib/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -132,6 +132,66 @@ sub find_by_recipient {
+    );
+
+    $self->sql->do(<<~"SQL", @args);
+        INSERT INTO edit_note_change (
+                        status, edit_note, changing_editor,
+                        old_note, new_note, reason
+                    )
+             VALUES (
+                        'edited', ?, ?,
+                        ?, ?, ?
+                    );
```
The spacing/indentation here seems kind of excessive somehow :)  Is this what the WIP SQL style guide we had suggested?

reosarevok replied:
Not exactly, but that might be worse still - it says "Subqueries and other parentheses-enclosed blocks should be aligned as though the open parenthesis were the 0-column. So, they should be indented as a unit, to identify them as a block. Subqueries should continue to have the opening keywords right-aligned." so `status` should be further right than the parens in `edit_note_change`...

yvanzo replied:
I don’t find it excessive. I wouldn’t mind aligning on parenthesis’ next column either, but it’s fine to me as-is for now.

mwiencek replied:
Fine as-is for me too.

reosarevok commented on lib/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -132,6 +132,35 @@ sub find_by_recipient {
+                    )
+             VALUES (
+                        'deleted', ?, ?,
+                        ?, '', ?
```
I tested the suggestion by adding an actual `Data::EditNote` test for this, which was something I probably should have added to begin with.

reosarevok commented:
@brainzbot, retest this please

reosarevok commented:
@brainzbot, retest this please

mwiencek reviewed (changes requested):
Will continue testing this today, but posting my pending comments for now.

reosarevok commented on admin/sql/CreateTables.sql:
```diff
@@ -575,6 +575,18 @@ CREATE TABLE edit_note
+    id                  SERIAL, -- PK
+    status              edit_note_status,
+    edit_note           INTEGER NOT NULL, -- references edit_note.id
+    changing_editor 	INTEGER NOT NULL, -- references editor.id
```
Not a bad point with the consistency. I'll make the change (it's just the name, so I can't see any reason why it would cause problems).

reosarevok commented on root/edit/components/EditNote.js:
```diff
@@ -78,6 +128,21 @@ const EditNote = ({
+            <a
+              className="remove-item icon"
+              href={`/edit-note/${editNote.id}/delete`}
+              title={l('Remove edit note')}
```
We use Remove in the UI (for entities and whatnot) a lot more than we use Delete, so maybe I should just use Remove everywhere.

reosarevok commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,225 @@
+            component_props => {
+                title    => l('Can’t modify edit note'),
+                message  => l(
+                  'You are not allowed to enter or modify edit notes.',
```
Hmm, dunno. I guess we could reuse the same string and actually indicate "modify" there?

mwiencek replied:
Sure, I wouldn't mind adding "or modify" there. Otherwise it seems a little strange to have two different strings that share the same purpose.

mwiencek commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,225 @@
+    my $is_note_deleted = $edit_note->{latest_change} &&
+                          $edit_note->{latest_change}->{status} eq 'deleted';
+
+    if ($is_note_deleted && !$is_admin) {
```
Sorry if you mentioned this already, but is there a reason admins can delete notes multiples times? To update the "reason" I guess?

reosarevok replied:
Yes. The main purpose is so that, for example, if the editor deleted the note with a reason of "Fuck you all, assholes", an admin could "re-delete it" and change the reason. It seemed easier than implementing a third form to edit the reason.

reosarevok commented on lib/MusicBrainz/Server/Controller/EditNote.pm:
```diff
@@ -0,0 +1,225 @@
+        if ($all_edit_notes->[$i]->editor_id != $edit_note->editor_id) {
+            $has_reply = 1;
+            last;
+        }
```
That's actually a very good idea - implemented it in a separate commit.

reosarevok commented on lib/MusicBrainz/Server/Data/EditNoteChange.pm:
```diff
@@ -0,0 +1,58 @@
+    return 'MusicBrainz::Server::Entity::EditNoteChange';
+}
+
+sub get_latest
+{
+    my ($self, $id) = @_;
+    my $query = 'SELECT ' . $self->_columns .
+                ' FROM ' . $self->_table .
+                ' WHERE edit_note = ?' .
+                ' ORDER BY change_time DESC, id DESC LIMIT 1';
+    my $row = $self->sql->select_single_row_hash($query, $id)
+        or return undef;
+    return $self->_new_from_row($row);
+}
```
Should we make a ticket for this? :)

mwiencek replied:
Did that in 8fb9aecc94289e4d0d5b3b0b84baed323e992f56.

reosarevok commented:
@brainzbot, retest this please

yvanzo reviewed (commented):
A few weird things spotted by testing manually:
1. “Remove edit note” just removes its content, while the edit note metadata (author, timestamp) will stay. It is always possible (for admins at least) to change its content again. Maybe the form should be rather named “Clear edit note content” or “Blank edit note out”?
2. It’s not possible to submit the form to change edit note with a blank edit note. It should just results in removing/blanking it out instead. It’s a minor UI/UX bug.
3. It’s not possible to cancel the form to change edit note when removed/blanked out already. It’s a minor UI bug.
4. It’s not possible for an editor to change their own edit note if they “removed” it themself by mistake. It’s a minor UX bug.
5. It’s not possible to change an edit note even when all further edit notes by other editors have been removed/blanked out. It should be allowed. It’s a minor UX bug.
6. Admins have no way to see the change history of edit notes. It’s really missing to be able to check old edit notes by reported editors but it can probably be added in the next release after schema change.

reosarevok commented:
Things I disagree with: 
2) is intentional - it makes it harder to blank an edit note by mistake. The alternative would be to stop differentiating edit and removal at all, but I personally think having a distinction makes sense.
4) is also intentional - you cannot really blank an edit note by mistake, unless you don't read at all and blindly click confirm. By then, you might as well have to post a new note honestly.

Things I agree should be changed:
5) is super minor (how often is this going to happen in real life?), but it's a fair point - we could treat already deleted notes as the equivalent of modbot/own notes. This IMO only works if we do *not* allow 4 though - once a note is deleted, if someone else under it goes back and edits the note to "undelete it", the whole thing would get ridiculous. That said, given how unlikely the case is, I don't see a need to rush to add it now and it could just be added post-release.
6) is a very fair point, and would be a good tool for an admin. I agree with adding it post-release (it's not a trivial amount of work) but I'll do my best to work on it soon :) 

Other:
1) I think given the note itself is gone it's not unfair nor unclear to say it removes the note :) But I wouldn't mind renaming it to something else if you and @mwiencek can agree on a better option.
3) I'm not sure I follow what you mean :) What are the exact steps here?

yvanzo commented:
If `2` and `4` are intentional, I’ve not seen it mentioned anywhere, it’s not tested either.

`5` is not that unlikely:

- A: _1234 5678 9012 3456_
- B: _is that your credit card number?_
- A: _no that's yours, please “remove” your edit note so I can erase it_

Or:

- A: _azertyu_
- B: _*** yourself_
- admin “removes” the injurious B's edit note

I don’t see any conflict with `4` either.

For `1` the edit note author and date will still be shown, and the content can even be changed by the editor if `4`.

Steps to reproduce `3`:
- Open the form to change an edit note
- Blank out the field “New edit note” if not empty already
- Click on “Cancel”

mwiencek commented:
I cherry-picked the actual schema change commit to the `schema-change-2023-q2` branch, since that part seems settled at least, and it will make testing easier. So only the UI changes are left here.

reosarevok commented:
@yvanzo suggested we could mention when editing an edit note that this is meant for small changes and will not send an email nor actively notify the other editors/voters in any way of the change.

reosarevok commented:
Changes so far:
* Added a test for 4) - editor can't modify deleted edit notes.
* Added a formatting helper for edit note editing in the same way as we have for edit note adding. The help message there also indicates that emails will not be sent.

reosarevok commented:
Now I added an extra commit which gets rid of the `required`ness of the text field when modifying notes, and instead validates it with the edit note validation methods we already use for adding required notes. This is more clear regarding 2) and additionally fixes 3) by skipping validation when Cancel is pressed.

reosarevok commented:
@brainzbot, retest this please

yvanzo commented:
Class naming is pretty wild. The controller should probably be named `ModifyEditNote` instead of `EditNoteEdit`. It would better match the corresponding test class `Edit::ModifyNote` too.

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'since somebody else has already replied',
+        'Can’t delete note with replies',
```
To follow the verbiage in the UI:
```suggestion
        'Can’t remove note with replies',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'can’t change other users',
+        'Can’t delete note by other user',
```
Same as above:
```suggestion
        'Can’t remove note by other user',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'it was entered too long ago',
+        'Can’t delete too old note',
```
Same as above:
```suggestion
        'Can’t remove too old note',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'Are you sure you want to remove the following edit note',
+        'Can delete last note',
```
Same as above:
```suggestion
        'Can remove last note',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'This edit note was removed by its author. Reason given: “I did a dumb”.',
+        'Deletion message (with reason) is present',
```
Same as above:
```suggestion
        'Removal message (with reason) is present',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'This note has already been removed',
+        'Can’t delete already deleted note',
```
Same as above:
```suggestion
        'Can’t remove already removed note',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'This note has already been removed',
+        'Can’t modify already deleted note',
```
Same as above:
```suggestion
        'Can’t modify already removed note',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    is(
+        scalar @change_class_matches,
+        1,
+        'Only 1 set of change note controls shown to normal user after delete',
```
Same as above, plus a few details:
```suggestion
        'Only 1 set of change edit note controls shown to normal user after removing edit note',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    is(
+        scalar @change_class_matches,
+        6,
+        '6 sets of change note controls shown to admin',
```
Same as above:
```suggestion
        '6 sets of change edit note controls shown to admin',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'Are you sure you want to remove the following edit note',
+        'Admin can delete older notes',
```
Same as above above:
```suggestion
        'Admin can remove older notes',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'Are you sure you want to remove the following edit note',
+        'Admin can "delete" deleted notes to change reason',
```
Same as above, plus a bit of clarification:
```suggestion
        'Admin can remove again already removed notes to change reason',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Controller/Edit/ChangeNote.pm:
```diff
@@ -0,0 +1,254 @@
+    html_ok($mech->content);
+    $mech->content_contains(
+        'This edit note was removed by an admin. Reason given: “Editor made a mistake”.',
+        'Deletion message (with new reason) is present',
```
Same as above:
```suggestion
        'Removal message (with new reason) is present',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -161,6 +122,94 @@ like($email2_body, qr{This is my note!}, 'Email body has correct edit note text'
+    my $edit_data = MusicBrainz::Server::Data::Edit->new(c => $test->c);
+    my $en_data = MusicBrainz::Server::Data::EditNote->new(c => $test->c);
+
+    note('We delete the first edit note in edit 1');
```
Same as above, plus small improvements:
```suggestion
    note('We remove the first edit note to edit #1');
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -161,6 +122,94 @@ like($email2_body, qr{This is my note!}, 'Email body has correct edit note text'
+    my $row = $test->c->sql->select_single_row_hash(
+        'SELECT * FROM edit_note_change WHERE edit_note = 1',
+    );
+    is($row->{status}, 'deleted', 'The change is marked as a deletion');
```
Same as above:
```suggestion
    is($row->{status}, 'deleted', 'The change is marked as a removal');
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -161,6 +122,94 @@ like($email2_body, qr{This is my note!}, 'Email body has correct edit note text'
+    my $edit_data = MusicBrainz::Server::Data::Edit->new(c => $test->c);
+    my $en_data = MusicBrainz::Server::Data::EditNote->new(c => $test->c);
+
+    note('We change the text of the first edit note in edit 1');
```
Same as above above, reuse wording from the rest of the tests:
```suggestion
    note('We modify the note text for the first edit note to edit #1');
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -161,6 +122,94 @@ like($email2_body, qr{This is my note!}, 'Email body has correct edit note text'
+    is(
+        $edit->edit_notes->[0]->text,
+        'Platypus',
+        'The note text for the first note has been changed',
```
Same as always:
```suggestion
        'The note text for the first edit note has been modified',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -161,6 +122,94 @@ like($email2_body, qr{This is my note!}, 'Email body has correct edit note text'
+    );
+    is(
+        $edit->edit_notes->[1]->text,
+        'This is a later note',
+        'The note text for the second note is unchanged',
```
Same as always above:
```suggestion
        'This is a later edit note',
        'The note text for the second edit note is unmodified',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -161,6 +122,94 @@ like($email2_body, qr{This is my note!}, 'Email body has correct edit note text'
+    my $row = $test->c->sql->select_single_row_hash(
+        'SELECT * FROM edit_note_change WHERE edit_note = 1',
+    );
+    is($row->{status}, 'edited', 'The change is marked as an edit');
```
Same as same:
```suggestion
    is($row->{status}, 'edited', 'The change is marked as a modification');
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -161,6 +122,94 @@ like($email2_body, qr{This is my note!}, 'Email body has correct edit note text'
+    is(
+        $row->{old_note},
+        'This is a note',
+        'The old note is stored correctly',
```
Same:
```suggestion
        'The old edit note is stored correctly',
```

yvanzo commented on t/lib/t/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -161,6 +122,94 @@ like($email2_body, qr{This is my note!}, 'Email body has correct edit note text'
+    is(
+        $row->{new_note},
+        'Platypus',
+        'The new note is stored correctly',
```
Sss:
```suggestion
        'The new edit note is stored correctly',
```

yvanzo reviewed (commented):
Spotted one actual bug due to the refactoring. The rest are just innocuous oversights.

reosarevok commented on lib/MusicBrainz/Server/Data/EditNote.pm:
```diff
@@ -132,6 +132,62 @@ sub find_by_recipient {
+                        old_note, new_note, reason
+                    )
+             VALUES (
+                        'edited', ?, ?,
```
Yeah, that's what I was generally trying to follow here, but also yes, we already have edited / deleted hardcoded for the DB level.

reosarevok commented on root/edit/NotesReceived.js:
```diff
@@ -13,14 +13,12 @@ import PaginatedResults from '../components/PaginatedResults.js';
 import getRequestCookie from '../utility/getRequestCookie.mjs';
 
-import EditNote from './components/EditNote.js';
+import EditNoteEntry from './components/EditNoteEntry.js';
```
But I'm using it in DeleteNote / ModifyNote, which are not lists :)

yvanzo replied:
It’s a list of 1 element :)
The current name didn’t help me much to understand what this component is compared to `EditNote`.

yvanzo replied:
I just picked the closest name that makes some sense and is already in use.
`Entry` is too generic to make any sense out of context.
I thought it was about database entry.

It's always open to alternative suggestions.

reosarevok replied:
I don't have any better ideas, so I used yours :)

reosarevok commented on root/edit/components/EditNote.js:
```diff
@@ -69,6 +78,50 @@ const EditNote = ({
+  const isRecent = Boolean(
+    noteDate && (new Date().getTime() - noteDate.getTime()) < twentyFourHours,
+  );
+  const canBeChangedByOwner = isCurrentEditor && !hasReply &&
+                              isRecent && !isDeleted &&
+                              !isAddingNotesDisabled(user);
```
So you would want us to send this all the way up from Perl? We could, but it's only used for calculating whether to show edit controls for this note, and it seems like a lot more effort to pass a map of edit note id => boolean all the way?

reosarevok replied:
Then do please add a ticket, ideally with your implementation suggestion but not the end of the world if you just document the issue :)

yvanzo replied:
Added https://tickets.metabrainz.org/browse/MBS-13086 :)

yvanzo reviewed (approved):
I’m lost with numbers but it’s overdue now and safe for others’ edit notes.
